File: dbxml/examples/cxx/misc/debug.cpp
/*
 * See the file LICENSE for redistribution information.
 *
 * Copyright (c) 2004,2009 Oracle. All rights reserved.
 *
 *
 *******
 *
 * Debug API
 * A very simple Berkeley DB XML program that demonstrates
 * debug API usage.
 *
 * This program demonstrates how to set a custom debug listener
 * class on the XmlQueryContext, and how to derive a class form
 * XmlDebugListener. The derived class, MyDebugListener, simply
 * prints out the execution information when a query is executed.
 *
 */

#include <dbxml/DbXml.hpp>
#include <iostream>

using namespace DbXml;
using namespace std;

class MyDebugListener : public XmlDebugListener
{
public:
	MyDebugListener(const string query)
		: depth_(-1),
		  query_(query)
	{}

	void start(const XmlStackFrame *stack);
	void end(const XmlStackFrame *stack);
	void enter(const XmlStackFrame *stack);
	void exit(const XmlStackFrame *stack);
	void error(const XmlException error, const XmlStackFrame *stack);

private:
	int depth_;
	string query_;
	void report(string stage, const XmlStackFrame *stack);
};

void MyDebugListener::report(string stage, const XmlStackFrame *stack)
{
	// Function return indentation spaces
	string depthStr("");
	for(int i=0; i<depth_; i++)
		depthStr += string("    ");

	// Get the query expression by line and column
	// Cut the string starting from column number to the the end of the line
	int line = stack->getQueryLine();
	int column = stack->getQueryColumn();

	string::size_type start=0, end=0;
	for (int i=0; i<line ; i++) {
		start = end;
		end = query_.find_first_of('\n', start+1);
		if (end == start+1 || end == string::npos)
			end = query_.length();
	}
	end--;
	if (start != 0)
		start++;

	string exprStr = query_.substr(start+column-1, (end+1)-(start+column-1));

	// Output debug report
	cout << depthStr
	     << stage << ":\""
	     << exprStr << "\" at "
	     << stack->getQueryFile() << ":"
	     << line << ":"
	     << column << endl;
}

void MyDebugListener::start(const XmlStackFrame *stack)
{
	depth_++;
	report("{START}", stack);
}

void MyDebugListener::end(const XmlStackFrame *stack)
{
	report("{END}", stack);
	depth_--;
}

void MyDebugListener::enter(const XmlStackFrame *stack)
{
	depth_++;
	report("[ENTER]", stack);
}

void MyDebugListener::exit(const XmlStackFrame *stack)
{
	report("[EXIT]", stack);
	depth_--;
}

void MyDebugListener::error(const XmlException error, const XmlStackFrame *stack)
{
	cout << "An error occurred in expression at line:" << stack->getQueryLine()
	     << " column:" << stack->getQueryColumn() <<  endl;
	throw error;
}

int main(int argc, char **argv)
{
	string containerName = "debugContainer.dbxml";
	string query = string() +
		       "for $a in collection('debugContainer.dbxml')/root/a\n" +
		       "where $a>'2'\n" +
		       "return $a";
	string docContent = "<root><a>1</a><a>2</a><a>3</a><a>4</a></root>";

	try {
		// Create an XmlManager
		XmlManager mgr;

		// Because the containers will exist on disk, remove them
		// first if they exist
		if (mgr.existsContainer(containerName) != 0)
			mgr.removeContainer(containerName);

		//create container
		XmlContainer cont = mgr.createContainer(containerName);
		XmlUpdateContext uc = mgr.createUpdateContext();

		XmlIndexSpecification is = cont.getIndexSpecification();
		is.setAutoIndexing(false);
		cont.setIndexSpecification(is, uc);

		XmlDocument doc = mgr.createDocument();
		doc.setContent(docContent);
		doc.setName("test.xml");
		cont.putDocument(doc, uc);

		// Set the custom debug listener on XmlQueryContext
		MyDebugListener dl(query);
		XmlQueryContext qc = mgr.createQueryContext();
		qc.setDebugListener(&dl);

		cout << query << endl << endl
		     << "Execution process:" << endl << endl;

		XmlQueryExpression expr = mgr.prepare(query, qc);
		XmlResults re = expr.execute(qc);

	} catch (XmlException &xe) {
		cout << "XmlException: " << xe.what() << endl;
	}
	return 0;
}

File: dbxml/examples/java/misc/Debug.java
/*
 * See the file LICENSE for redistribution information.
 *
 * Copyright (c) 2004,2009 Oracle. All rights reserved.
 *
 *
 *******
 *
 * Debug API
 * A very simple Berkeley DB XML program that demonstrates
 * debug API usage.
 * 
 * This program demonstrates how to set a custom debug listener 
 * class on the XmlQueryContext, and how to derive a class form 
 * XmlDebugListener. The derived class, MyDebugListener, simply 
 * prints out the execution information when a query is executed.
 *   
 */

package misc;

import com.sleepycat.dbxml.*;

class MyDebugListener extends XmlDebugListener {

	private int depth_ = -1;
	private String outputFormat = "%s%s:\"%s\" at %s:%s:%s\n";
	private String query_;
	
	public MyDebugListener(String query){
		query_ = query;
	}
	
	public void start(XmlStackFrame stack) throws XmlException {
		depth_++;
		System.out.printf(outputFormat, getIndentation(depth_), "{START}", getExprStr(stack), stack.getQueryFile(), stack.getQueryLine(),stack.getQueryColumn());
	}

	public void end(XmlStackFrame stack) throws XmlException {
		System.out.printf(outputFormat, getIndentation(depth_), "{END}", getExprStr(stack), stack.getQueryFile(), stack.getQueryLine(),stack.getQueryColumn());
		depth_--;
	}

	public void enter(XmlStackFrame stack) throws XmlException {
		depth_++;
		System.out.printf(outputFormat, getIndentation(depth_), "[ENTER]", getExprStr(stack), stack.getQueryFile(), stack.getQueryLine(),stack.getQueryColumn());
	}

	public void exit(XmlStackFrame stack) throws XmlException {
		System.out.printf(outputFormat, getIndentation(depth_), "[EXIT]", getExprStr(stack), stack.getQueryFile(), stack.getQueryLine(),stack.getQueryColumn());
		depth_--;
	}

	public void error(XmlException error, XmlStackFrame stack)
			throws XmlException {
		System.out.printf("An error occurred in expression. line:%s column:%s\n", stack.getQueryLine(),stack.getQueryColumn());
        throw error;
	}

	// Function return indentation spaces 	
	private String getIndentation(int depth) {
		StringBuffer sb = new StringBuffer("");
		
		for(int i=0; i<depth; i++)
			sb.append("    ");
		
		return sb.toString();
	}
	
	// Get the query expression by line and column
	// Cut the string starting form column number to the 
	// end of the line 
	private String getExprStr(XmlStackFrame stack) throws XmlException{

		int line = stack.getQueryLine();
		int column = stack.getQueryColumn();
		
        int start=0, end=0;

        for (int i=0; i < line ; i++){
                start = end;          
                end = query_.indexOf("\n", start+1);
                if (end == start+1 || end == -1)              
                        end = query_.length();   
        }                                        

        end--;

        if (start != 0)
                start++;

        return query_.substring( start+column-1, end+1 );

	}
	
}


public class Debug {

	public static void main(String[] args) throws Throwable {
		
		String containerName = "debugContainer.dbxml";
		String docName = "test.xml";
		String docString = "<root><a>1</a><a>2</a><a>3</a><a>4</a></root>";
		
		String query = "for $a in collection('debugContainer.dbxml')/root/a\n"
		 	 			+ "where $a>'2'\n" 
		 	 			+ "return $a";
		
		XmlManager mgr = new XmlManager();
		XmlContainer cont = null;
		XmlResults re = null;
		XmlQueryExpression expr = null;
		
		try {

			// Because the container will exist on disk, remove it
            // first if exist
			if (mgr.existsContainer(containerName) != 0)
				mgr.removeContainer(containerName);
			
			//create container
			cont = mgr.createContainer(containerName);
			XmlIndexSpecification is = cont.getIndexSpecification();
			is.setAutoIndexing(false);
			cont.setIndexSpecification(is);

			XmlUpdateContext uc = mgr.createUpdateContext();
			cont.putDocument(docName, docString, uc);

			// Set the custom debug listener on XmlQueryContext
			MyDebugListener dl = new MyDebugListener(query);
			XmlQueryContext qc = mgr.createQueryContext();
			qc.setEvaluationType(XmlQueryContext.Eager);
			qc.setDebugListener(dl);
			
			expr = mgr.prepare(query, qc);
			System.out.println("Query: ");
			System.out.println(query);
			System.out.println("\nExecution process:\n");
			
			re = expr.execute(qc);

		} catch (XmlException e) {
			System.out.println("Exception: " + e.getMessage());
		} finally {
			// Clean up
			expr.delete();
			re.delete();			
			cont.delete();
			mgr.delete();
        }
	}

}

